src/JustSaying/Messaging/Channels/IChannelConsumer.cs

```diff
@@ -0,0 +1,45 @@
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using JustSaying.AwsTools.MessageHandling;
+using JustSaying.Messaging.Monitoring;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Channels
+{
+    internal interface IChannelConsumer
+    {
+        Task Start();
+        IChannelConsumer ConsumeFrom(IAsyncEnumerable<IQueueMessageContext> messageSource);
+    }
+
+    internal class ChannelConsumer : IChannelConsumer
+    {
+        private IAsyncEnumerable<IQueueMessageContext> _messageSource;
+        private readonly IMessageDispatcher _dispatcher;
+        private readonly ILogger _logger;
+
+        public ChannelConsumer(IMessageDispatcher dispatcher, ILoggerFactory logger)
+        {
+            _dispatcher = dispatcher;
+            _logger = logger.CreateLogger<IChannelConsumer>();
+        }
+
+        public IChannelConsumer ConsumeFrom(IAsyncEnumerable<IQueueMessageContext> messageSource)
+        {
+            _messageSource = messageSource;
+            return this;
+        }
+
+        public async Task Start()
+        {
+            await foreach (var messageContext in _messageSource)
```

Comment:
As Martin said, take a cancellation token and change this to `_messageSource.WithCancellation(cancellationToken)`

Reply:
See my comment about cancellation above 🙂

Reply:
Done 🙂